internal/jimm/model_cleanup.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 Canonical.
+
+package jimm
+
+import (
+	"context"
+	"fmt"
+
+	jujuparams "github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/state"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+)
+
+// CleanupModelsDying loops over dying models, contacting the respective controller.
+// And deleting the model from our database if the error is `NotFound` which means the model was successfully deleted.
+func (j *JIMM) CleanupModelsDying(ctx context.Context) error {
+	const op = errors.Op("jimm.CleanupModelsDying")
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	err := j.DB().ForEachModel(ctx, func(m *dbmodel.Model) error {
+		if m.Life == state.Dying.String() {
+			// if the model is dying and not found by querying the controller we can assume it is dead.
+			// And safely delete the reference from our db.
+			api, err := j.dialModel(ctx, &m.Controller, m.ResourceTag())
+			if err != nil {
```

Comment:
can you still dial a model if it doesn't exist? wouldn't we expect an error here where we try to dial it?

Reply:
yeah, i think it too. I switched to dialing the controller.
Now I am thinking, if a controller is deleted without being removed from jimm, the dangling models will be there forever.
Shouldn't be an issue right?

Reply:
hmm.. we should deal with controller removal in a separate pr

Reply:
yes, i can do that